src/lib/accessibility/FocusZone/FocusTrapZone.tsx

```diff
@@ -0,0 +1,322 @@
+  implements IFocusTrapZone {
+  private static _focusStack: FocusTrapZone[] = []
+  private _root: { current: HTMLElement | null } = { current: null }
+  private _previouslyFocusedElementOutsideTrapZone: HTMLElement
+  private _previouslyFocusedElementInTrapZone?: HTMLElement
+  private windowElement: Window | null
+
+  private createRef = elem => (this._root.current = ReactDOM.findDOMNode(elem) as HTMLElement)
+
+  static propTypes = {
+    as: customPropTypes.as,
+    componentRef: PropTypes.object,
+    className: PropTypes.string,
+    elementToFocusOnDismiss: PropTypes.object,
+    ariaLabelledBy: PropTypes.string,
+    isClickableOutsideFocusTrap: PropTypes.bool,
+    ignoreExternalFocusing: PropTypes.bool,
+    forceFocusInsideTrap: PropTypes.bool,
+    firstFocusableSelector: PropTypes.string,
+    disableFirstFocus: PropTypes.bool,
+    focusPreviouslyFocusedInnerElement: PropTypes.bool,
+  }
+
+  static handledProps = [
+    'as',
+    'className',
+    'componentRef',
+    'elementToFocusOnDismiss',
+    'ariaLabelledBy',
+    'isClickableOutsideFocusTrap',
+    'ignoreExternalFocusing',
+    'forceFocusInsideTrap',
+    'firstFocusableSelector',
+    'disableFirstFocus',
+    'focusPreviouslyFocusedInnerElement',
+  ]
+
+  static defaultProps: IFocusTrapZoneProps = {
+    as: 'div',
+    isClickableOutsideFocusTrap: true,
+  }
+
+  public componentDidMount(): void {
+    FocusTrapZone._focusStack.push(this)
+    const { disableFirstFocus = false } = this.props
+
+    this.windowElement = getWindow(this._root.current)
+    this._previouslyFocusedElementOutsideTrapZone = this._getPreviouslyFocusedElementOutsideTrapZone()
+
+    if (
+      !this._root.current.contains(this._previouslyFocusedElementOutsideTrapZone) &&
+      !disableFirstFocus
+    ) {
+      this.focus()
+    }
+
+    this._hideContentFromAccessibilityTree()
+    this._subscribeToEvents()
+  }
+
+  public render(): JSX.Element {
+    const { className, ariaLabelledBy } = this.props
+    const rest = getUnhandledProps({ handledProps: FocusTrapZone.handledProps }, this.props)
+    const ElementType = getElementType({ defaultProps: FocusTrapZone.defaultProps }, this.props)
+
+    return (
+      <ElementType
+        {...rest}
+        className={className}
+        ref={this.createRef}
+        aria-labelledby={ariaLabelledBy}
+        onKeyDown={this._onKeyboardHandler}
+        onFocusCapture={this._onFocusCapture}
+      >
+        {this.props.children}
+      </ElementType>
+    )
+  }
+
+  public componentWillUnmount(): void {
+    const { ignoreExternalFocusing } = this.props
+
+    FocusTrapZone._focusStack = FocusTrapZone._focusStack.filter((value: FocusTrapZone) => {
+      return this !== value
+    })
+
+    const activeElement = document.activeElement as HTMLElement
+    if (
+      !ignoreExternalFocusing &&
+      this._previouslyFocusedElementOutsideTrapZone &&
+      typeof this._previouslyFocusedElementOutsideTrapZone.focus === 'function' &&
```

Comment:
which condition (semantically) we would like to check with this expression?

Reply:
that is defined and has focus function.
I think this can be enough
```this._previouslyFocusedElementOutsideTrapZone &&
      this._previouslyFocusedElementOutsideTrapZone.focus
```
What do you think?

Reply:
IMHO, the name would be better ```isFocusable```. Cause ```hasFocus``` it's more about a state, not ability. But I now think about skipping this check. As we never done this across app, to keep consistency.